src/components/buildinfo.ts

```diff
@@ -0,0 +1,868 @@
+                                ctx.font = 0.8 * this.graph.resolutionMultiplier + 'rem serif';
+                                ctx.fillText(
+                                    'ms',
+                                    this.graph.margins.left - this.graph.textMargin * this.graph.resolutionMultiplier,
+                                    closestPoint.y + this.rem * this.graph.resolutionMultiplier * 0.8
+                                );
+
+                                ctx.globalAlpha = 0.7;
+                                ctx.font = 1.2 * this.graph.resolutionMultiplier + 'rem serif';
+                                ctx.textAlign = 'center';
+                                ctx.textBaseline = 'top';
+                                ctx.fillText(
+                                    closestPoint.runName.replace(/(\\d+)\\-(\\w+)/, '$2 (rule $1)'),
+                                    (ctx.width + this.graph.margins.left - this.graph.margins.right) / 2,
+                                    this.graph.margins.top
+                                );
+                            } else {
+                                this.drawGraph();
+                            }
+                        }
+                    };
+                </script>
+            </body>
+        </html>
+        `
+        // #endregion
+
+        if (this.currentBuild) {
+            this.panel.reveal(vscode.ViewColumn.Beside)
+            this.panel.webview.postMessage({
+                type: 'init',
+                startTime: this.currentBuild.buildStart,
+                stepTimes: this.currentBuild.stepTimes,
+                pageTotal: this.currentBuild.pageTotal
+            })
+        }
+    }
+
+    private displayProgress(current: string | number) {
+        if (!this.currentBuild) {
+            throw Error(`Can't Display Progress for non-Started build - see BuildInfo.buildStarted()`)
+        }
+
+        this.configuration = vscode.workspace.getConfiguration('latex-workshop')
+
+        if (current === 0) {
+            this.currentBuild.stepTimes[`${this.currentBuild.ruleNumber}-${this.currentBuild.ruleName}`][`T${+new Date()}-Wait Time`] =
+                +new Date() - this.currentBuild.lastStepTime
+        } else {
+            if (this.currentBuild.ruleProducesPages) {
+                // if page already exists, add times instead of making new entry
+                const pageAlreadyExistsRegex = new RegExp(`^T\d+-PAGE:${current}`)
+                const pageMatchArray = Object.keys(this.currentBuild.stepTimes[`${this.currentBuild.ruleNumber}-${this.currentBuild.ruleName}`]).map(pageLabel => Boolean(pageLabel.match(pageAlreadyExistsRegex)));
+                if (pageMatchArray.indexOf(true) !== -1) {
+                    this.currentBuild.stepTimes[`${this.currentBuild.ruleNumber}-${this.currentBuild.ruleName}`][Object.keys(this.currentBuild.stepTimes)[pageMatchArray.indexOf(true)]] += +new Date() - this.currentBuild.lastStepTime
```

Comment:
`Object.keys(this.currentBuild.stepTimes)[pageMatchArray.indexOf(true)]` needs to be ``Object.keys(this.currentBuild.stepTimes[`${this.currentBuild.ruleNumber}-${this.currentBuild.ruleName}`])[pageMatchArray.indexOf(true)]``

This fixes both the undefined entry and the entry for previous rules.

Reply:
Done